proposed/MPID Messaging.md

```diff
@@ -0,0 +1,451 @@
+    1. Client-to-Vault request to take ownership of the Vault
+    1. Vault-to-Client notification of low resources on the Vault
+
+    In this case, the existing library infrastructure would probably need significant changes to allow a Vault to act as an MPID Client (e.g. the MPID struct is defined in the SAFE Client library).
+
+- Another point is that (as with MAID accounts), there is no cleanup done by the network of MpidMessages if the user decides to stop using SAFE.
+
+- Also, not exactly within the scope of this RFC, but related to it; MPID packets at the moment have no human-readable name.  It would be more user-friendly to provide this functionality.
+
+# Appendix
+
+### Further Implementation Details
+
+All MPID-related messages will be in the form of a Put, Post or Delete of a `StructuredData`.
+
+This will be:
+
+```rust
+
+StructuredData {
+    type_tag: MPID_MESSAGE,
+    identifier: mpid_message_name(mpid_message),  // or mpid_header_name(signed_header)
+    data: XXX,
+    previous_owner_keys: vec![],
+    version: 0,
+    current_owner_keys: vec![sender_public_key],
+    previous_owner_signatures: vec![],
+}
+
+```
+
+where XXX is indicated in the following list of all MPID-related messages:
+
+Sent from Client:
+
+| Message | From ==> To |
+|:---|:---|
+| Post[Online] |           Either Client    ==> Own Managers |
+| Put[Message] |           Sender Client    ==> Own Managers |
+| Post[Has[Vec<Header>]] | Sender Client    ==> Own Managers |
+| Post[GetAllHeaders] |    Sender Client    ==> Own Managers |
+| Delete[Header] |         Either Client    ==> Own Managers (try to delete from inbox and outbox) |
+| Delete[Header] |         Recipient Client ==> Sender's Managers |
+
+
+Sent from Vault:
+
+| Message | From ==> To |
+|:---|:---|
+| PutResponse[Error[Message]] |              Sender's Managers (outbox)   ==> Sender Client |
+| Put[Header] |                              Sender's Managers (outbox)   ==> Recipient's Managers (inbox) |
+| PutResponse[Error[Header]] |               Recipient's Managers (inbox) ==> Sender's Managers (outbox) |
+| Post[GetMessage[Header]] |                 Recipient's Managers (inbox) ==> Sender's Managers (outbox) |
+| PostResponse[Error[GetMessage[Header]]] |  Sender's Managers (outbox)   ==> Recipient's Managers (inbox) |
+| Post[Message] |                            Sender's Managers (outbox)   ==> Recipient's Managers (inbox) |
+| Post[Message] |                            Recipient's Managers (inbox) ==> Recipient Client |
+| Post[HasResponse[Vec<Header>]] |           Sender's Managers (outbox)   ==> Sender Client |
+| Post[GetAllHeadersResponse[Vec<Header>]] | Sender's Managers (outbox)   ==> Sender Client |
+
+
+The various different types for `StructuredData::data` can be enumerated as:
+
+```rust
+#[allow(variant_size_differences)]
+enum MpidMessageWrapper {
+    /// Notification that the MPID Client has just connected to the network
+    Online,
```

Comment:
I don't think a separate Online message needs to be presented.
A `GetAllHeaders` shall do the job and it does make more sense: when a client is interested in what's available in his Inbox, then there is meaning to push to it.